### PR TITLE
Update the build pipeline

### PR DESCRIPTION
**Pull request by dmint789**
This is a big PR, but this was a very important update for us to make. It should improve the developer experience greatly when it comes to building the PDFs to see a preview of the documents, which was previously only really easy to do by creating a PR and waiting for the Github pipeline to do it, which always took a while and made it harder to iterate with minor adjustments. This PR does the following:

- Update the build script to use weasyprint instead of wkhtmltopdf, which is no longer maintained and causes bugs in the output PDFs sometimes. Some other minor improvements to this script have also been made.
- Add a new build script that uses Docker for consistent file rendering.
- Unify the duplicated styles between the documents and edudoc stylesheets, and keep the directory-specific styles in their own separate files: documents-style.css and edudoc-style.css
- Make minor adjustments to the documents themselves to ensure there are no alignment issues and to make some of the Markdown look cleaner.
- Renumber points 5. and 6. of the [WCA Bylaws](https://documents.worldcubeassociation.org/documents/WCA%20Bylaws.pdf) section 2.4 as 4. and 5. respectively (point 4. was missing). Keep in mind that it's displayed as 1-5 in production rn, because Markdown seems to ignore what number you actually put there and simply uses the correct one in order.

### CODE REVIEW

File: bin/docker_build.sh
Comment (dmint789) [reply bin/docker_build.sh:1]: Fixed

File: README.md
Comment (dmint789) [reply README.md:1]: But you left a review comment. Should I just resolve it then?

**Comment by dmint789** [comment]
@nsilvestri could you provide an example of an unwanted page break due to the spacing between lines? For me the spacing looks the same as it does in production rn.

**Comment by dmint789** [comment]
Staff crash course fixed

**Comment by nsilvestri** [comment]
@dmint789 the page breaks caused by spacing are not "unwanted", just slightly different. 1-2 lines will be pushed to the next page, or vice versa. Very minor stylistic thing, just noting it as an inconsistency.

**Comment by dmint789** [comment]
Fixed the bullet point issue in the competitor tutorial

**Comment by dmint789** [comment]
Okay, now I need to update the Github actions

**Comment by dmint789** [comment]
Found another issue: the counters reset when there is content between two items (see the competition templates doc for an example). Currently working on fixing this.

**Comment by dmint789** [comment]
Fixed

**Comment by dmint789** [comment]
I have personally gone through every single markdown document in the repository, fixed all alignment issues and made sure they all look almost exactly the same as the original docs rendered with wkhtmltopdf. The PR is now ready for review.

@nsilvestri could you please review the changes in the documents and edudoc folders? It's almost exclusively alignment fixes and changes from pure HTML to markdown/pandoc syntax in some of our old docs. My expectation is that you won't find any issues, as I tried being as thorough as possible, but it's good to have another set of eyes on the changes.

@thewca/software-team could you please review the changes in the root of the repo? Namely, the changes to the assets and bin folders, the README and my newly-created Dockerfile.

I will also review all of the aforementioned changes myself for good measure.

**Comment by dmint789** [comment]
@thewca/software-team also, I'll need someone to override the PR build test, cause it's failing due to not being up to date with these changes

**Review by nsilvestri — APPROVED** [review]
Thanks for seeing through this colossal effort. LGTM on behalf of WQAC.

File: Dockerfile
Comment (gregorbg) [inline Dockerfile:15]: Out of curiosity: The above `WORKDIR /home/app` path kinda implies that there is already a user called `app` pre-defined by the Ubuntu image. Does that one have the necessary privileges?

File: Dockerfile
Comment (gregorbg) [inline Dockerfile:1]: Have you considered using something like `alpine` instead?
Downloading a whole Ubuntu image for a couple of PDFs feels a bit overkill.

File: assets/edudoc-style.css
Comment (gregorbg) [inline assets/edudoc-style.css:25]: Where do the variables like `headerTitle`, `page`, etc. come from?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:4]: Why did you decide to drop this?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:15]: There is a "delete" step at the very bottom of the build script, and because of that, individually removing PDFs here wasn't necessary before. Why is it necessary now?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:20]: You dropped the part that prints information about the stylesheet being used. This could make debugging harder.

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:41]: Am I correct in reading that you're basically pre-pending `stdin` to the existing file?

This `echo` is hard to read and understand, and manually typing out `\n` line breaks is never a good idea in scripts. If this actually does what I think it does, then consider some other approach along the lines of:
1. Piping the `sed` stuff into a new. temporary file using `echo > $temp_file`
2. Appending the original file using `cat $file >> $temp_file`
3. Renaming `$temp_file` to the original `$file` if necessary.

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:47]: What is this `sed` doing?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:55]: The two cases are completely identical as far as I can tell, so there's no need for an if-else.

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:64]: In line with my other comment above (feel free to answer only one of them): Why did you move this `rm` step that is clearly present in the left side of the diff into the conversion method body?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:81]: If you need to echo a line break at the end of the string, something with your previous output is wrong. Never manually encode line breaks in script files, previous `echo` prints always terminate with a new line.

If some `weasyprint` commands don't terminate with a new line and you really really need this extra line break, then issue two `echo` statements with the first `echo`ing an empty string. (And probably raise an issue against WeasyPdf, terminal output commands should always be terminated with line breaks for readability.)

File: bin/docker_build.sh
Comment (gregorbg) [inline bin/docker_build.sh:23]: This is what the `||` control flow operator in Bash is made for `try_some_command --arg "foo" || echo 'the command failed!'`. In this context, you probably want `docker build || exit`.

Comment (gregorbg) [reply bin/docker_build.sh:23]: https://linuxsimply.com/bash-scripting-tutorial/redirection-and-piping/piping/double-pipe

File: bin/install_dependencies.sh
Comment (gregorbg) [inline bin/install_dependencies.sh:8]: The reason why we did not use `apt` before is that it allows us to pin the specific version of `pandoc` and `wkhtmltopdf`. That way, unexpected breaking changes can be avoided.
Your new `apt` way looks more convenient on the surface, but actually it will always just take the newest version that is available at the time of running the script. In six months or one year, that might be a new major version that suddenly needs new CLI args to be passed, or otherwise the output looks horrible. (Like, say for example the CLI parameter for defining "use this CSS file please" changes from `--css $some_css_file` to `--stylesheet $some_css_file`. In the "best" case scenario, the old `weasyprint` invocation will fail because then `--css` becomes an unrecognized CLI option. In the "worst" case, it will just continue, silently skipping the `--css` option, which means our documents get compiled without stylesheet and "suddenly" look awful and everybody will scratch their heads what happened because these issues are hard to debug.

Comment (gregorbg) [reply bin/install_dependencies.sh:8]: TL;DR: Is there a way for you to pin specific versions of at least `pandoc` and `weasyprint`?
(The font files don't need pinned versions because we don't invoke CLI commands on them that could break in the future)

**Review by gregorbg — CHANGES_REQUESTED** [review]
This review includes the technical scripts only. I have not reviewed every single `md` file nor have I meticulously verified the generated PDF contents. That is left entirely up to WQAC.

File: Dockerfile
Comment (dmint789) [reply Dockerfile:1]: I have, but that uses a different package manager, so the install script cannot be used with it. This PR used debian slim before, but I figured I might as well make this image as close to the github action as possible. If you think Debian slim is close enough to ubuntu, I can change it back.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:4]: I felt that it was unnecessary, seeing as the docker build pipeline just handles everything for you anyways. This was kinda used for debugging for you.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:15]: I found myself sometimes waiting for a specific file to build, and this made it easier to spot it when it wasn't buried in .md and .html files, so I think this is a little better.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:20]: You're right, and since two stylesheets can be used now, I could update this to show which stylesheets are being used.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:41]: That's exactly how it used to work. I felt like the temporary file was unnecessary, so I changed it. I don't think this part is that bad personally.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:47]: Yeah, so as the comment above it mentions, there is this totally unnecessary styling that pandoc generates that's not supported by weasyprint, and this sed just removes it. I have not found this affecting anything, it just removes the annoying warnings during the build process.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:64]: Answered above

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:55]: Oops, good catch. There used to be a difference with wkhtmltopdf.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:81]: Nono, it totally does have a line break before anyways, this just adds another line break to visually separate this line from the stuff above it.

Comment (dmint789) [reply bin/build.sh:81]: If you feel strongly about it, I can change this to two echo statements

File: bin/docker_build.sh
Comment (dmint789) [reply bin/docker_build.sh:23]: Ooh, thanks for sharing! Didn't know about that one

File: bin/install_dependencies.sh
Comment (dmint789) [reply bin/install_dependencies.sh:8]: Good point there. I think the best option would be to specify the package versions using apt. I'll make this change.

File: Dockerfile
Comment (dmint789) [reply Dockerfile:15]: This is the output I get when doing ls -l /home/app:

`drwxr-xr-x 11 ubuntu ubuntu 4096 May 28 10:06 wca-documents`

File: assets/edudoc-style.css
Comment (dmint789) [reply assets/edudoc-style.css:25]: These are just custom names I chose, and they're used below inside of the `@media print` block

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:41]: Oh, but the new line characters are totally unnecessary, I'll get rid of them

File: Dockerfile
Comment (dmint789) [reply Dockerfile:1]: But actually, I didn't find a big difference in build times between the two, and you only need to build the image once and then you can keep using it, so I'd just leave Ubuntu as the base image

Comment (gregorbg) [reply Dockerfile:1]: Okay yeah then whatever. It's not about build times necessarily, more about how much space you're taking up on the host machine.

File: bin/build.sh
Comment (gregorbg) [reply bin/build.sh:78]: Hm, in that case I guess it's fine.

File: bin/build.sh
Comment (gregorbg) [reply bin/build.sh]: I still do feel quite strongly that `echo $(cat /dev/stdin)` is not a good idiom, especially when piping. Not sure what actual Bash gurus say about this, but unless you find a really compelling argument or an article that underlines that the way you're doing it is common best practice, I'd rather have it be separate echo statments like before.

File: Dockerfile
Comment (gregorbg) [reply Dockerfile:15]: That does not fully answer my question. Does the user `app` actually exist or am I misreading some file/folder configuration?
If the user exists, please try out whether it has the necessary privileges for running wasyprint. I cannot figure this out by staring at `chmod`esque permission return values.

Comment (dmint789) [reply Dockerfile:15]: Oh, my bad, I misunderstood. The app user doesn't exist in the ubuntu docker image, no. This directory can simply be used for your app's files.

File: Dockerfile
Comment (dmint789) [reply Dockerfile:1]: Right. So the image with an Ubuntu base is 1.76GB. I just tried the Debian slim base image again, but it won't work, because of the newly-added package version locking for pandoc and weasyprint, since Debian and Ubuntu have slightly different versions of packages in their respective repos.

**Comment by dmint789** [comment]
@gregorbg is this good to be merged now?

File: bin/build.sh
Comment (gregorbg) [inline bin/build.sh:43]: I still feel quite strongly about this `echo` with a wrapped `cat /dev/stdin` call.
What's so hurtful about a plain, simple, readable

```bash
sed -E | sed -E | sed -E > $target_file
cat $origin_file >> $target_file
```

File: documents-guide.md
Comment (gregorbg) [inline documents-guide.md:103]: Might be worth bumping the version?

File: documents-guide.md
Comment (gregorbg) [inline documents-guide.md:1]: For what it's worth, you're changing quite a lot of inline HTML to custom pandoc `::::`-style syntax throughout all documents.

image

It might be worth explaining the usage of `::::: { .html-class }` and similar snippets here.

File: documents/minutes/2017-12-12 WCA Board Founding Meeting.md
Comment (gregorbg) [inline documents/minutes/2017-12-12 WCA Board Founding Meeting.md:23]: You're changing this to become a heading (MD `####` element) whereas previously it was just "some text" wrapped in a `b` tag.

Why do you also additionally need to make it bold with `**text**`? Shouldn't headings be bold by themselves?

Comment (gregorbg) [reply documents/minutes/2017-12-12 WCA Board Founding Meeting.md:23]: And if they are indeed not bold by default (which would be very surprising!) then have you considered making them bold as part of your custom CSS stylesheet?

File: bin/docker_build.sh
Comment (gregorbg) [inline bin/docker_build.sh:8]: What are the two different styles of calling this script, and why are they both necessary?

File: bin/docker_build.sh
Comment (gregorbg) [inline bin/docker_build.sh:17]: You can set a `$is_rebuilld` flag in the "Get the directory to build argument" block above, and then just do one "equal to true" check here

**Review by gregorbg — CHANGES_REQUESTED** [review]

File: documents-guide.md
Comment (dmint789) [reply documents-guide.md:103]: This is not the actual version of this document, it's an example of how you can use the version class. This doc doesn't have versioning, we can just update it like the README.

File: documents-guide.md
Comment (dmint789) [reply documents-guide.md:1]: It's explained on lines 31-39

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:43]: The problem is that we need to prepend here, not append. That's what the wrapped cat call achieves here.

Comment (dmint789) [reply bin/build.sh:43]: Found a different, hopefully cleaner, solution using tee

File: documents/minutes/2017-12-12 WCA Board Founding Meeting.md
Comment (dmint789) [reply documents/minutes/2017-12-12 WCA Board Founding Meeting.md:23]: They're bold by default, that's my bad

File: bin/docker_build.sh
Comment (dmint789) [reply bin/docker_build.sh:8]: You can do one of the following: script, script --rebuild, script edudoc, script documents, script edudoc --rebuild, script documents --rebuild, script --rebuild edudoc, script --rebuild documents.

File: bin/docker_build.sh
Comment (dmint789) [reply bin/docker_build.sh:17]: But that one doesn't check for the flag being == to --rebuild, it checks the opposite along with another check. I'd have to add more if statements to do what you're suggesting. Would you like me to do that?

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh]: That actually didn't work, and it breaks every other document. Reverting it to the one line echo solution.

Comment (dmint789) [reply bin/build.sh]: Let me know if you figure out a cleaner way to prepend the header html

Comment (gregorbg) [reply bin/build.sh:44]: Uhm... Doesn't my original suggestion do just that?
Run some `sed`s, write them to a new, temporary file (i.e. they get added to the *beginning* of that new file) and then use `>>` to _append_ the original file to the `sed`'ed stuff.

File: bin/docker_build.sh
Comment (gregorbg) [reply bin/docker_build.sh:17]: Okay, I didn't consider the mumbo-jumbo'd way of appending the `--rebuild` argument in odd places. Nevermind,

File: bin/docker_build.sh
Comment (gregorbg) [reply bin/docker_build.sh:8]: Okay, in that case it should be fine as-is.

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh:44]: Ohh, so you meant to append to the new file that has the header html in it. I'm dumb, I'll do that then.

File: bin/docker_build.sh
Comment (dmint789) [reply bin/docker_build.sh:17]: I figured it would be easier for the user if the order of the arguments didn't matter

File: bin/build.sh
Comment (dmint789) [reply bin/build.sh]: Done

**Comment by dmint789** [comment]
Looks like all issues are resolved now

**Review by gregorbg — APPROVED** [review]
The technical build scripts have my approval. I did not (and will not) check the final rendered output of every single document in the new pipeline.

Sidenote: The CI error log looks like you forgot `sudo` somewhere

**Comment by dmint789** [comment]
Oh, maybe it's the apt commands. Also, does the build pipeline use the setup files from master or from this branch I'm trying to merge? I assumed the CI error would never go away until this PR is merged.

**Comment by dmint789** [comment]
@thewca/board can you please approve? This probably requires Board approval, because I had to edit many of the documents themselves, but the only real change is the last point in the description of this PR, but it doesn't even affect the rendered file, since the numbers themselves are ignored. I just wanted to fix the numbering in the Bylaws Markdown. I have personally checked the rendering of every single document in this repo, and everything has remained consistent, bar some very minor alignment here and there.

**Comment by dmint789** [comment]
@gregorbg Looks like I've answered my own question, as the PR build test has a different output after my commit now. I'll try to fix it, looks like the repos used by the github action are different from the ubuntu image repos.

**Comment by dmint789** [comment]
@gregorbg according to [this](https://itsfoss.com/apt-install-specific-version/), package version locking is not possible on ubuntu. I'd just remove the pandoc and weasyprint package versions and call it a day, unless you know of a way to get a specific version of these packages from apt.

**Review by nsilvestri — DISMISSED** [review]
Approved on behalf as Board regarding non-content or formatting changes to Board-sensitive documents.

**Comment by dmint789** [comment]
@gregorbg the checks are finally passing. Can you please re-approve, cause the new commits made your approval stale. Maybe we can revisit package version locking in another issue and PR, cause this is currently blocking two of our projects.

**Review by gregorbg — APPROVED** [review]

**Review by nsilvestri — DISMISSED** [review]

**Pull request merged**
